src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheckTest.java

```diff
@@ -300,4 +300,65 @@ public void testClassOnly() throws Exception {
                 expected);
     }
 
+    @Test
+    public void testLineSeparationBeforeComments() throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(EmptyLineSeparatorCheck.class);
+        checkConfig.addAttribute("allowMultipleEmptyLines", "false");
+        final String[] expected = {
+            "19: " + getCheckMessage(MSG_SHOULD_BE_SEPARATED, "package"),
+            "23: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "27: " + getCheckMessage(MSG_MULTIPLE_LINES, "import"),
+            "32: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "39: " + getCheckMessage(MSG_MULTIPLE_LINES, "/*"),
+            "50: " + getCheckMessage(MSG_MULTIPLE_LINES, "/*"),
+            "67: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "78: " + getCheckMessage(MSG_MULTIPLE_LINES, "/*"),
+            "83: " + getCheckMessage(MSG_MULTIPLE_LINES, "import"),
+            "89: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "93: " + getCheckMessage(MSG_MULTIPLE_LINES, "/*"),
+            "101: " + getCheckMessage(MSG_MULTIPLE_LINES, "VARIABLE_DEF"),
+            "106: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "113: " + getCheckMessage(MSG_MULTIPLE_LINES, "/*"),
+            "126: " + getCheckMessage(MSG_MULTIPLE_LINES, "/*"),
+            "139: " + getCheckMessage(MSG_MULTIPLE_LINES, "METHOD_DEF"),
+            "146: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "156: " + getCheckMessage(MSG_MULTIPLE_LINES, "/*"),
+            "171: " + getCheckMessage(MSG_MULTIPLE_LINES, "/*"),
+            "188: " + getCheckMessage(MSG_MULTIPLE_LINES, "CLASS_DEF"),
+            "194: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "198: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "204: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "216: " + getCheckMessage(MSG_MULTIPLE_LINES, "/*"),
+            "229: " + getCheckMessage(MSG_MULTIPLE_LINES, "/*"),
+            "243: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "246: " + getCheckMessage(MSG_SHOULD_BE_SEPARATED, "INTERFACE_DEF"),
+            "251: " + getCheckMessage(MSG_MULTIPLE_LINES, "/*"),
+            "266: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "275: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "288: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "293: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "299: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "307: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "316: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "322: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "342: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+            "350: " + getCheckMessage(MSG_MULTIPLE_LINES, "//"),
+        };
+        verify(checkConfig,
+                getPath("InputEmptyLineSeparatorWithComments.java"),
```

Comment:
Please add test cases of comments you ignore to show there will be no violations.

Reply:
added, please review

Reply:
> Checks for empty line separators after header, package, all import declarations, fields, constructors, methods, nested classes, static initializers and instance initializers.

@kazachka Maybe I am missing it, but I don't see examples of like comments inside constructors/methods/lambdas followed by many empty lines. What about comments at the end outside the top level class?

Reply:
added more comments, please review

Reply:
Perfect thanks.